C-API: undefined reference to `wasmer_last_error_length()'

<!-- Thanks for the bug report! -->

### Describe the bug

Using the C-API on x64 Linux on Version 1.0.0-beta2, i can't use the C-API functions to retrieve compilation errors.

### Steps to reproduce

I've simply copy pasted the error-fetching code that is in the comments of wasmer_wasm.h and tried to compile it. These are the only functions that don't seem to link.

### Expected behavior

No linking error :D

I'm using Wasmer using CMake's external project using the release tar.gz: 

```
include(ExternalProject)

# Get wasmer
ExternalProject_Add(
    wasmer_src
    PREFIX "vendor/wasmer"  
    URL https://github.com/wasmerio/wasmer/releases/download/1.0.0-beta2/wasmer-linux-amd64.tar.gz
    BUILD_COMMAND ""
    INSTALL_COMMAND ""
    CONFIGURE_COMMAND ""
    CMAKE_COMMAND ""
    UPDATE_COMMAND ""
    BUILD_BYPRODUCTS <SOURCE_DIR>/lib/libwasmer.a
)

# Prepare wasmer
ExternalProject_Get_Property(wasmer_src source_dir)
set(wasmer_INCLUDE_DIR ${source_dir}/include)
set(wasmer_LIBRARY_PATH ${source_dir}/lib/libwasmer.a)
file(MAKE_DIRECTORY ${wasmer_INCLUDE_DIR})
add_library(wasmer STATIC IMPORTED)
set_property(TARGET wasmer PROPERTY IMPORTED_LOCATION ${wasmer_LIBRARY_PATH})
set_property(TARGET wasmer APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES ${wasmer_INCLUDE_DIR})

# Dependencies
add_dependencies(wasmer wasmer_src)
```

and then in the simplest example possible:

```cxx
#include <wasmer_wasm.h>

int main() {
    int len = wasmer_last_error_length();
    return 0;
}
```


compiled with (by CMake) `/bin/c++ -g -O0 -fsanitize=address  wasmer-test.cc.o  -o wasmertest libwasmer.a  -pthread`

results in 

```
[build] /usr/bin/ld: wasmer-test.cc.o: in function `main':
[build] wasmer-test.cc:4: undefined reference to `wasmer_last_error_length()'
```

The library is defining `wasmer_last_error_length` and `wasmer_last_error_message` as `nm` says they are, the problem is that those aren't the names the C++ code is looking for. From the error, 
```
[build] wasmer-test.cc:4: undefined reference to `wasmer_last_error_length()'
```
that's actually a symbol named `_Z24wasmer_last_error_lengthv` being pretty-printed by running it through `c++filt _Z24wasmer_last_error_lengthv`. If it were just the symbol name, it wouldn't have the empty parentheses on the end because for names that aren't mangled in C++, we don't know their parameters.

wasmer_wasm.h needs to wrap all its declarations in an `extern "C"{ /* ... */ }` block.

--- a/lib/c-api/build.rs
+++ b/lib/c-api/build.rs
@@ -240,6 +240,7 @@ fn new_builder(language: Language, crate_dir: &str, include_guard: &str, header:
     Builder::new()
         .with_config(cbindgen::Config {
             sort_by: cbindgen::SortKey::Name,
+            cpp_compat: true,
             ..cbindgen::Config::default()
         })
         .with_language(language)

--- a/lib/c-api/wasmer.h
+++ b/lib/c-api/wasmer.h
@@ -55,7 +55,11 @@
 #include <stdlib.h>
 
 #if defined(WASMER_WASI_ENABLED)
-enum Version {
+enum Version
+#ifdef __cplusplus
+  : uint8_t
+#endif // __cplusplus
+ {
 #if defined(WASMER_WASI_ENABLED)
   /**
    * Version cannot be detected or is unknown.
@@ -82,13 +86,19 @@ enum Version {
   Snapshot1 = 3,
 #endif
 };
+#ifndef __cplusplus
 typedef uint8_t Version;
+#endif // __cplusplus
 #endif
 
 /**
  * List of export/import kinds.
  */
-enum wasmer_import_export_kind {
+enum wasmer_import_export_kind
+#ifdef __cplusplus
+  : uint32_t
+#endif // __cplusplus
+ {
   /**
    * The export/import is a function.
    */
@@ -106,7 +116,9 @@ enum wasmer_import_export_kind {
    */
   WASM_TABLE = 3,
 };
+#ifndef __cplusplus
 typedef uint32_t wasmer_import_export_kind;
+#endif // __cplusplus
 
 /**
  * The `wasmer_result_t` enum is a type that represents either a
@@ -128,7 +140,11 @@ typedef enum {
  *
  * See `wasmer_value_t` to get a complete example.
  */
-enum wasmer_value_tag {
+enum wasmer_value_tag
+#ifdef __cplusplus
+  : uint32_t
+#endif // __cplusplus
+ {
   /**
    * Represents the `i32` WebAssembly type.
    */
@@ -146,7 +162,9 @@ enum wasmer_value_tag {
    */
   WASM_F64,
 };
+#ifndef __cplusplus
 typedef uint32_t wasmer_value_tag;
+#endif // __cplusplus
 
 typedef struct {
 
@@ -408,6 +426,10 @@ typedef struct {
   wasmer_byte_array host_file_path;
 } wasmer_wasi_map_dir_entry_t;
 #endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
 
 /**
  * Creates a new Module from the given wasm bytes.
@@ -1578,5 +1600,9 @@ wasmer_import_object_t *wasmer_wasi_generate_import_object_for_version(unsigned 
  */
 Version wasmer_wasi_get_version(const wasmer_module_t *module);
 #endif
+
+#ifdef __cplusplus
+} // extern "C"
+#endif // __cplusplus
 
 #endif /* WASMER_H */

--- a/lib/c-api/wasmer_wasm.h
+++ b/lib/c-api/wasmer_wasm.h
@@ -124,6 +124,10 @@ typedef struct wasi_env_t wasi_env_t;
 typedef struct wasi_version_t wasi_version_t;
 #endif
 
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
+
 #if defined(WASMER_WASI_ENABLED)
 void wasi_config_arg(wasi_config_t *config, const char *arg);
 #endif
@@ -566,5 +570,9 @@ const char *wasmer_version_pre(void);
  * See the module's documentation.
  */
 void wat2wasm(const wasm_byte_vec_t *wat, wasm_byte_vec_t *out);
+
+#ifdef __cplusplus
+} // extern "C"
+#endif // __cplusplus
 
 #endif /* WASMER_WASM_H */

